hook/hooktest/service.go

```diff
@@ -0,0 +1,212 @@
+package hooktest
+
+import (
+	"sync"
+
+	"gopkg.in/errgo.v1"
+
+	"github.com/juju/gocharm/charmbits/service"
+	"github.com/juju/gocharm/hook"
+)
+
+// NewServiceFunc returns a function that can be used to create
+// new services and can be assigned to service.NewService
+// to mock the OS-level service creator.
+// When the service is started, the given Runner
+// will be used to run it (via hook.Main).
+//
+// If the notify channel is not nil, it will be used to send
+// events about services created with the function.
+// It should be buffered with a size of at least 2.
+func NewServiceFunc(r *Runner, notify chan ServiceEvent) func(service.OSServiceParams) service.OSService {
+	services := &osServices{
+		notify:   notify,
+		runner:   r,
+		services: make(map[string]*osService),
+	}
+	return func(p service.OSServiceParams) service.OSService {
+		services.mu.Lock()
+		defer services.mu.Unlock()
+		if svc := services.services[p.Name]; svc != nil {
+			return svc
+		}
+		svc := &osService{
+			params:   p,
+			services: services,
+		}
+		services.services[p.Name] = svc
+		return svc
+	}
+}
+
+type osServices struct {
+	notify chan ServiceEvent
+	runner *Runner
+
+	// mu guards the following fields and services.
+	mu       sync.Mutex
+	services map[string]*osService
+}
+
+// osService provides a mock implementation of the
+// service.OSService interface.
+type osService struct {
+	params service.OSServiceParams
+
+	services *osServices
+
+	// The following fields are guarded by services.mu.
```

Comment:
No, this is still relevant. We're using the wider-scoped osServices mutex to guard the
fields in all the individual services too.